### ISSUE TITLE
Global optimize algorithm gets outside of boundaries parameters

### ISSUE THREAD

**Issue by Axoahi** [ljvmiranda921/pyswarms#315]
**Describe the bug**
We are trying to optimize a DBSCAN algorithms parameters, and it works nice with a few iterations and particles, but when we increase them, we get the error "eps must be positive" and its boundaries are between 0.001 -- 5

**To Reproduce**
Execute a DBSCAN algorithm

**Expected behavior**
It should finish the search :)

**Environment (please complete the following information):**
 - OS: Windows 10
 - Version Education Version 1803
 - PySwarms Version 1.0.0
 - Python Version 3.6.5

**Additional context**
We have done few tests, and globaloptimizer works properly with 1000 iterations and 50 particles, but crashes with 2000 iterations and 100 particles
We also have detected with optimizer.history NaN values when it crashes

**Comment by Axoahi** [comment]
#### DBSCAN Optimization
#### Parameters PSO ######
iterationsDBSCAN = 2000
particlesDBSCAN = 100
# Range of values for DBSCAN paramters in PSO Optimization
# eps = 0.001 -- 5
minEps = 0.001
maxEps = 0.5
# Samples = 5 -- 30
minSamples = 5
maxSamples = 30
######################


#Create bounds up and down for the parameters of optimize function
max_bound = maxSamples * np.ones(2)
max_bound[0] = maxEps
min_bound = minEps * np.ones(2)
min_bound[1] = minSamples
limite = (min_bound, max_bound)

# instatiate the optimizer(parameters for the swarms(weight to the communication of the swarms(best swarms, best local swarms and inertia)))
options = {'c1': 0.5, 'c2': 0.3, 'w': 2}
# #Create swarm with n_particles, dimensions of dimensional space, parameters for the swarms and bounds
optimizer = GlobalBestPSO(n_particles=particlesDBSCAN, dimensions=2, options=options, bounds=limite)

cost, pos = optimizer.optimize(f_dbscan_accuracy, iters=iterationsDBSCAN, fast=True)

**Comment by ljvmiranda921** [comment]
Hi @Axoahi we'll get back to you shortly! 
Wanna take this on, if you have time? @whzup @stevenbw ?

**Comment by whzup** [comment]
Hey @Axoahi, thanks for your question. 😄
This error message is not from pyswarms and the `optimizer.hist` attribute only [stores the results of your objective function](https://github.com/ljvmiranda921/pyswarms/blob/master/pyswarms/single/global_best.py#L186). To me it seems like it is a problem with your objective function and not with `pyswarms`. Could you please check if `f_dbscan_accuracy` is able to return `NaN` values?

**Comment by Axoahi** [comment]
Hi guys, thanks for your attention and your job.
@whzup, It could be possible, but it depends of input parameters, that global optimizer sends. The point is, is it possible to work fine with a set of parameters, and work bad with others? Maybe I have minderstood something :P
Anyway, here is the code of the two other functions that I am using, and a img with the error
If you need anything else, no problem, i will provide it

```
def f_dbscan_accuracy(x):
    n_particles = x.shape[0]
    j = [optimize_dbscan_accuracy(x[i]) for i in range(n_particles)]
    #j = [forward_prop(x[i]) for i in range(n_particles)]
    return np.array(j)

def optimize_dbscan_accuracy(x):
    #Train Algorithm
    dbscan = DBSCAN(algorithm='auto', eps=x[0], min_samples=x[1])
    #Test model
    prediction = dbscan.fit_predict(X_norm)
    #Get labels with ground truth
    truevalue = y
    #Apply Accuracy_score(true label, predicted label)
    score = accuracy_score(truevalue, prediction)
    #Apply Accuracy_score to prediction of SVM
    return (1-score)
```
![svm-pso](https://user-images.githubusercontent.com/12016601/53566444-7b1c6800-3b5c-11e9-8994-8017314ab65d.jpg)

**Comment by stevenbw** [comment]
Hi @Axoahi , are you able to show us the `x` parameter that you used? That is, the parameter which you are passing into the `f_dbscan_accuracy` function.

**Comment by Axoahi** [comment]
Hi @stevenbw, I am using the tutorial from the docs

https://pyswarms.readthedocs.io/en/latest/examples/custom_objective_function.html

In that tutorial, the function f(x) (the equivalent to mine f_dbscan_accuracy). As long as i know, the goal of that function is to move the particles in the dimensional space and then calls the objective-main function (in my case, optimize-dbscan-accuracy)

If you ask me, I would say that the parameters that you send is a tuple, formed by two valors inside the boundaries that I have provided. For instance, I have printed some x values inside the f_dbscan_accuracy function, and they are:

[0.35087949 9.04448519]
[ 0.17525069 10.10525934]
[0.35087949 9.04448519]
[ 0.17525069 10.10525934]

First one is inside eps boundaries (0.001 -- 5) and the second one is inside minSamples (5 -- 30)

**Comment by stevenbw** [comment]
Hi @Axoahi thanks for the information.

I have been able to replicate the bug and added a print statement to print the particle positions. 

It seems that one particle's `min_sample` (`x[1]`) position goes from 25 to `NaN` causing the error from DBSCAN. Could this be caused by a `NaN` velocity @whzup ?

**Comment by whzup** [comment]
@stevenbw You can easily check if any velocities are `NaN`. They should be stored in the `hist` attribute as well.

I won't have a lot of time during the weekend so I'd propose that we should try to use `pdb` to debug this one. I think it should be easier to find the error when you can go through the program step by step. Maybe this [SO question](https://stackoverflow.com/questions/242485/starting-python-debugger-automatically-on-error) is useful for that. Also, I'm not familiar with the DBSCAN algorithm could you provide the whole code you try to run @Axoahi?

**Comment by Axoahi** [comment]
Hi @stevenbw 
Sure, here you have the whole code for the execution, and the data that you need for creating the dataframe. You know, you must put the csv file on same directory as python code

```
import pandas as pd
import numpy as np
import time as tm
from sklearn import decomposition
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score
from pyswarms.single.global_best import GlobalBestPSO

X_norm = pd.read_csv("normData.csv", sep=",")
y = X_norm["34"]

def f_dbscan_accuracy(x):
    print(x[0])
    print(x[1])
    n_particles = x.shape[0]
    j = [optimize_dbscan_accuracy(x[i]) for i in range(n_particles)]
    return np.array(j)

def optimize_dbscan_accuracy(x):
    #Train Algorithm
    dbscan = DBSCAN(algorithm='auto', eps=x[0], min_samples=x[1])
    #Test model
    prediction = dbscan.fit_predict(X_norm)
    #Get labels with ground truth
    truevalue = y
    #Apply Accuracy_score(true label, predicted label)
    score = accuracy_score(truevalue, prediction)
    #Apply Accuracy_score to prediction of SVM
    return (1-score)

%%time
from pyswarms.single.global_best import GlobalBestPSO
from sklearn.cluster import DBSCAN
from sklearn import preprocessing
from sklearn.metrics import accuracy_score
from sklearn import metrics
from sklearn.metrics import pairwise_distances

#### DBSCAN Optimization
#### Parameters PSO ######
iterationsDBSCAN = 2000
particlesDBSCAN = 100
# Range of values for DBSCAN paramters in PSO Optimization
# eps = 0.001 -- 5
minEps = 0.001
maxEps = 0.5
# Samples = 5 -- 30
minSamples = 5
maxSamples = 30
######################


#Create bounds up and down for the parameters of optimize function
max_bound = maxSamples * np.ones(2)
max_bound[0] = maxEps
min_bound = minEps * np.ones(2)
min_bound[1] = minSamples
limite = (min_bound, max_bound)

# instatiate the optimizer(parameters for the swarms(weight to the communication of the swarms(best swarms, best local swarms and inertia)))
options = {'c1': 0.5, 'c2': 0.3, 'w': 2}
# #Create swarm with n_particles, dimensions of dimensional space, parameters for the swarms and bounds
optimizer = GlobalBestPSO(n_particles=particlesDBSCAN, dimensions=2, options=options, bounds=limite)

# #Run the optimization
# cost, pos = optimizer.optimize(f_dbscan_silhouette, iters=iterationsDBSCAN, fast=True)

cost, pos = optimizer.optimize(f_dbscan_accuracy, iters=iterationsDBSCAN, fast=True)
```

[normData.zip](https://github.com/ljvmiranda921/pyswarms/files/2926270/normData.zip)

**Comment by Axoahi** [comment]
BTW, it takes around 8 hours to get a good solution, but it crashes around 4 hours after it gets started

**Comment by stevenbw** [comment]
Hi @Axoahi - sorry for the delay in responding. I should have some time over the next few days to look into this further. Thanks for providing all of the code and data. Does an eps value of 0.001 make sense in your domain? My first impression is that it is not related to the data but more likely to the accuracy function.

**Comment by Axoahi** [comment]
Hi @stevenbw. Take it easy, I posted this in order to improve pyswarms. And yes, an eps=0.001 make sense in my domain (its the minimal radius to clusterize our data, taking in account that our data has been normalized). What I am trying to do with pyswarms and dbscan, is to see which combination of eps and samples is the suitable to get the highgest accuracy

**Comment by ljvmiranda921** [comment]
@stalebot stop

**Comment by shouvik-paul** [comment]
Hi Everyone,
I am trying to implement k-means clustering using PSO (actually I want to implement k means clustering using PYSwarms). Primarily I want to test it on Iris dataset then others.
I am new.  Can anyone help me by providing me  a  basic knowledge of the uses of PySwarm?
or Can anyone help me by providing a demo code for 'k means clustering using PYSwarms'?

**Comment by whzup** [comment]
@shouvik-paul Hi! Your question might be better suited to be asked in the gitter channel https://gitter.im/pyswarms/Issues 😄 Just log in with your GitHub login.

**Comment by whzup** [comment]
@Axoahi Could you maybe provide a minimal working example for the issue you run into? I'll be honest no one of us maintainers has the time and energy to debug code that runs 4 hours before getting any results😕.

**Comment by shouvik-paul** [comment]
@whzup thank you for your reply. can you please tell me, How can I check Pbest or Gbest in the case of a dataset like wine dataset?

**Comment by Axoahi** [comment]
Hi @whzup. Sorry, but I cant reproduce the error faster or with a minimal solution. I guess that inertia parameters can make that particles get outside initial boundaries. I started the algorithm at night, and when i woke up, i saw the error 😐
